torchtune/models/gemma/_tokenizer.py

```diff
@@ -0,0 +1,129 @@
+        self,
+        path: str,
+    ):
+        self._spm_model = SentencePieceBaseTokenizer(path)
+
+        # Original tokenizer has no pad_id, which causes indexing errors when batch training
+        self._spm_model.pad_id = 0
+
+        # This is used in tokenize_messages: if the tokenizer does not
+        # encode whitespace, then we can more easily split strings
+        # on whitespace characters and encode them separately.
+        self.encodes_whitespace = any(
+            [self._spm_model.encode(c) for c in WHITESPACE_CHARS]
+        )
+
+        # During generation, stop when eos_id is encountered
+        self.stop_tokens = [self.eos_id]
+
+    @property
+    def eos_id(self):
+        return self._spm_model.eos_id
+
+    @property
+    def bos_id(self):
+        return self._spm_model.bos_id
+
+    @property
+    def pad_id(self):
+        return self._spm_model.pad_id
+
+    @property
+    def vocab_size(self):
+        return self._spm_model.vocab_size
+
+    def encode(
+        self,
+        text: str,
+        add_bos: bool = True,
+        add_eos: bool = True,
+        trim_leading_whitespace: bool = False,
+    ) -> List[int]:
+        return self._spm_model.encode(
+            text,
+            add_bos=add_bos,
+            add_eos=add_eos,
+            trim_leading_whitespace=trim_leading_whitespace,
+        )
+
+    def decode(
+        self,
+        token_ids: List[int],
+    ) -> str:
+        return self._spm_model.decode(token_ids)
+
+    def tokenize_messages(
+        self, messages: List[Message], max_seq_len: Optional[int] = None
+    ) -> Tuple[List[int], List[bool]]:
+        r"""Tokenize a list of messages one at a time then concatenate them,
+        returning a list of tokens and a list of masks.
+
+
+        Example:
+            >>> tokenizer = GemmaTokenizer(tokenizer_path)
+            >>> messages = [
+                Message(role="system", content="system message\n", masked=True),
+                Message(role="user", content="user prompt\n", masked=True),
+                Message(role="assistant", content="assistant response\n"),
+            ]
+            # tokenize_messages encodes messages separately and concats
+            >>> tokenizer.tokenize_messages(messages, max_seq_len)[0]
+            [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+
+            # Same result as encoding the full string in one go
+            >>> tokenizer.encode(''.join([message.content for message in messages]))
+            [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+
+        Args:
+            messages (List[Message]): A list of messages, each containing role, content,
+                and masked attributes.
+            max_seq_len (Optional[int]): A max sequence length to truncate tokens to.
+                Default: None
+
+        Returns:
+            Tuple[List[int], List[bool]]: The tokenized messages
+        """
+        return tokenize_messages_no_special_tokens(
```

Comment:
Gemma has special tokens, no?

Reply:
Yeah what's the latest on this? It seems to me that it does have special tokens: https://huggingface.co/google/gemma-2b/blob/main/tokenizer_config.json

Reply:
these aren't used in any capacity in our current gemma implementation. Also I think we discussed elsewhere that the gemma and mistral special token updates will be a follow up: #1118